### ISSUE TITLE
Cannot `freeze-to-file` and `thaw-from-file` objects that are larger than 2GB

### ISSUE THREAD

**Issue by RokLenarcic** [taoensso/nippy#172]
Hello. Due to these functions being merely a thin wrapper around `freeze` and `thaw` they interact with a file via a byte array. That limits the size of these operations to max size of byte array which is 2GB. I'd love to have the option to freeze and thaw to DataInput/DataOutput objects or ByteBuffer objects. 

Incidentally ByteBuffer offers very similar API than DataInput and DataOutput, and the benefit there is MappedByteBuffer use, which is what you get when you memory map a file.

Obviously such use cannot use the current encryptor and compressor system as those need a byte array. But I am sure there's some kind of streaming compression and encryption that could be used.

**Comment by ptaoussanis** [comment]
@RokLenarcic Hi Rok,

>  I'd love to have the option to freeze and thaw to DataInput/DataOutput objects or ByteBuffer objects.

Just checking if you noticed [`freeze-to-out!`](https://cljdoc.org/d/com.taoensso/nippy/3.4.0-RC3/api/taoensso.nippy#freeze-to-out!) and [`thaw-from-in!`](https://cljdoc.org/d/com.taoensso/nippy/3.4.0-RC3/api/taoensso.nippy#thaw-from-in!)?

> Incidentally ByteBuffer offers very similar API than DataInput and DataOutput, and the benefit there is MappedByteBuffer use, which is what you get when you memory map a file.

I'd be open to a sketch/proof-of-concept PR of some sort if you or someone else felt like proposing a concrete way to support ByteBuffers 👍

**Comment by RokLenarcic** [comment]
Yeah I saw those functions, but there's a rather large amount of logic (e.g. header and options parsing/use) that is in `freeze` that I would have to replicate in my own code to have `freeze-to-out!` work approx the same.

In general the direction that would be needed for large objects is to have a protocol that has functions like a stream: write bytes and close/finalize. Then you could have encryption, compression and the writing implementation itself implement this. Then you could stack these like we do with streams, encryption wrapping the compression wrapping the data/file writing instance of this protocol.

**Comment by ptaoussanis** [comment]
Sketch/proof-of-concept PR welcome if you or anyone else felt like taking a look at this - otherwise I can look at this myself for a future version of Nippy.

Please 👍 here if there's interest in adding better streaming support to Nippy.

**Comment by ourkwest** [comment]
As of 3.7, does this limit also affect "freeze-to-out!"? Either that or I'm doing it wrong.

**Comment by ptaoussanis** [comment]
> @ourkwest  As of 3.7, does this limit also affect "freeze-to-out!"? Either that or I'm doing it wrong.

Please see https://github.com/taoensso/nippy/releases#release-v3.7.1 which addresses an unintended size limit regression in v3.7, apologies.

**Comment by ptaoussanis** [comment]
Closing as part of issue triage. Please see `freeze-to-out!`, or `freeze-to-bb!` in v3.7 which should be helpful for most folks. Otherwise feel free to re-open if someone would like to add a PR to enhance the `freeze-to-file` and `thaw-from-file` utils 👍

**Comment by ourkwest** [comment]
@ptaoussanis Thank you for fixing this quickly. And no need for apologies - we're grateful for the free library!
